Billing diff

zanchi-r opened the pull request:
This PR adds two endpoints:

- The absolute endpoint that returns the absolute usagetype cost for each week/month in the time range
- The variations endpoint that returns the usagetype cost variations between each pair of week/month in the time range

Both these endpoints can return JSON or CSV depending on the content type specified by the user

zanchi-r committed:
routes: serve CSV requests

This adds the possibility to serve CSV content for routes that return data implementing the csvGenerator interface

zanchi-r committed:
diff: Added endpoints to get absolute diff data and variations

This adds two endpoint:
- The absolute endpoint that returns the absolute usagetype cost for each week/month in the time range
- The variations endpoint that returns the usagetype cost variations between each pair of week/month in the time range

zanchi-r committed:
diff: added tests

This adds tests for the es_request_constructor

schubev commented on routes/routes.go:
```diff
@@ -51,20 +53,38 @@ type Decorator interface {
+	contentType := r.Header["Content-Type"][0]
+	w.Header()["Content-Type"] = []string{fmt.Sprintf("%s; charset=utf-8", contentType)}
+	switch contentType {
+	case "application/json":
```
HTTP Content-Type negociation is more complicated than that : user agents can specify a list of accepted content types, with prioritization, and usage of wildcards. For example a client may request ```Accept: text/*;q=0.5 application/json;q=0.8``` where `application/json` should be selected. I think this should be implement as a multiplexer like `routes.MethodMuxer` and as a decorator.

schubev commented on costs/diff/diff_route.go:
```diff
@@ -0,0 +1,207 @@
+
+// diffQueryArgs allows to get required queryArgs params
+var diffQueryArgs = []routes.QueryArg{
+	routes.QueryArg{
```
You should reuse `aws.AwsAccountsOptionalQueryArgs` instead of redefining it.

schubev commented on costs/diff/diff_route.go:
```diff
@@ -0,0 +1,207 @@
+		Type:        routes.QueryArgStringSlice{},
+		Optional:    true,
+	},
+	routes.QueryArg{
```
You should try and reuse `QueryArg` instances.

schubev commented on costs/diff/diff_route.go:
```diff
@@ -0,0 +1,207 @@
+
+// validateAwsAccounts will validate awsAccounts passed to it.
+// It checks that they are numbers that are 12 character long
+func validateAwsAccounts(parsedParams esQueryParams) error {
```
Doesn’t this function already exist in the cost breakdown implementation ?

zanchi-r committed:
diff: added diff endpoint

This replaces the absolute and variation endpoints by the diff endpoint that combines all the data.
The variation is now a percentage to make it easier to understand

zanchi-r committed:
diff: reuse aws.AwsAccountsOptionalQueryArg instead of redefining it

zanchi-r committed:
diff: reuse aws account validation function

Moved the validateAwsAccounts in the aws package in order to share it in the rest of the code and avoid repetition

zanchi-r committed:
routes: Select a response Content-Type based on the Accept header

This handles a basic Accept header containing a single type without quality factor.

zanchi-r committed:
costs: fix accountList type

This replaces the accountList type from []string to []int for consistency

zanchi-r committed:
routes: reuse common QueryArgs to avoid repetition

This adds a commonQueryArgs file in the routes package to centralize the QueryArgs that are used in multiple places
